### PR TITLE
Move SSH tunnel processing to a background thread

### PR DESCRIPTION
**Pull request by beckerhe**
I expected that there is not much more to do than `service_deploy_manager->moveToThread(ssh_background_thread);`, but some cleanup needed to be done as well.

Furthermore some synchronization needed to happen to make the public interface of ServiceDeployManager thread-safe.

### CODE REVIEW

File: OrbitQt/servicedeploymanager.cpp [hunk 1]
```diff
@@ -199,6 +222,15 @@ outcome::result<void> ServiceDeployManager::CopyOrbitServicePackage() {
 
 ErrorMessageOr<void> ServiceDeployManager::CopyFileToLocal(std::string_view source,
                                                            std::string_view destination) {
+  if (QThread::currentThread() != thread()) {
+    ErrorMessageOr<void> result = outcome::success();
+    DeferToBackgroundThreadAndWait(
+        this, [&, source = std::string{source}, destination = std::string{destination}]() {
+          result = CopyFileToLocal(source, destination);
```
Comment (dimitry-) [inline OrbitQt/servicedeploymanager.cpp:229]: Maybe move this to another function and have a CHECK(QThread::currentThread() != thread()) there?

In which case you won't need this if here.

Comment (dimitry-) [reply OrbitQt/servicedeploymanager.cpp:229]: Same for Cancel

File: OrbitQt/servicedeploymanager.cpp [hunk 2]
```diff
@@ -384,6 +416,12 @@ void ServiceDeployManager::StartWatchdog() {
 }
 
 outcome::result<ServiceDeployManager::GrpcPort> ServiceDeployManager::Exec() {
+  if (QThread::currentThread() != thread()) {
+    std::optional<outcome::result<GrpcPort>> result;
+    DeferToBackgroundThreadAndWait(this, [&]() { result = Exec(); });
+    return result ? result.value() : outcome::failure(std::errc::timed_out);
+  }
```
Comment (dimitry-) [inline OrbitQt/servicedeploymanager.cpp:423]: and here

File: OrbitQt/servicedeploymanager.cpp [hunk 3]
```diff
@@ -478,6 +516,11 @@ void ServiceDeployManager::ShutdownSession() {
 }
 
 void ServiceDeployManager::Shutdown() {
+  if (QThread::currentThread() != thread()) {
+    DeferToBackgroundThreadAndWait(this, [this]() { Shutdown(); });
+    return;
+  }
```
Comment (dimitry-) [inline OrbitQt/servicedeploymanager.cpp:522]: and here

File: OrbitQt/servicedeploymanager.cpp (hunk 1 above)
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp]: I restructured this entirely and moved the thread into ServiceDeployManager which means it now always uses a background thread.

File: OrbitQt/servicedeploymanager.cpp [hunk 4]
```diff
@@ -384,6 +421,12 @@ void ServiceDeployManager::StartWatchdog() {
 }
 
 outcome::result<ServiceDeployManager::GrpcPort> ServiceDeployManager::Exec() {
+  outcome::result<GrpcPort> result = outcome::success(GrpcPort{0});
+  DeferToBackgroundThreadAndWait(this, [&]() { result = ExecImpl(); });
+  return result;
+}
+
+outcome::result<ServiceDeployManager::GrpcPort> ServiceDeployManager::ExecImpl() {
```
Comment (dimitry-) [inline OrbitQt/servicedeploymanager.cpp:429]: Do you want to add CHECKS here to make sure this is always called on background thread? just in case...

**Review by dimitry- — APPROVED** [review]
otherwise lgtm, thanks!

File: OrbitQt/servicedeploymanager.cpp (hunk 4 above)
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:429]: Yes. That's already done but not pushed yet. There is also another problem on Linux which needs to be fixed before merging that! Thanks.

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:455]: Linux is now also working. QEventLoop didn't like to be created in a different thread. By deferring its initialization to the background thread the problem is gone. Might be a bug or an insufficiently communicated precondition of the class.

**Comment by beckerhe** [comment]
It works now under Linux. (The last commit describes in detail what went wrong). Adding Pierric for review since he might wanna test this before merging anyway.

File: OrbitQt/servicedeploymanager.cpp [hunk 5]
```diff
@@ -94,6 +93,9 @@ ServiceDeployManager::ServiceDeployManager(const DeploymentConfiguration* deploy
 
   background_thread_.start();
   moveToThread(&background_thread_);
+  QMetaObject::invokeMethod(
```
Comment (dimitry-) [inline OrbitQt/servicedeploymanager.cpp:96]: maybe duplicate the explanation here as well? "Event loop needs to be initialized in background thread to work properly on Linux"?

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:96]: Yep, good idea. On it.

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:98]: done.

**Comment by pierricgimmig** [comment]
@beckerhe , I tried your change and it feels really robust now on Linux, I couldn't reproduce the hang. It seems like data is streaming more fluidly too. Thanks for the fix! I'll look at the code shortly.

File: OrbitQt/servicedeploymanager.cpp [hunk 6]
```diff
@@ -313,13 +375,14 @@ outcome::result<void> ServiceDeployManager::StartOrbitServicePrivileged() {
...
 
-  OUTCOME_TRY(loop_.exec());
+  OUTCOME_TRY(loop_->exec());
   QObject::connect(&orbit_service_task_.value(), &OrbitSshQt::Task::errorOccurred, this,
                    &ServiceDeployManager::handleSocketError);
   return outcome::success();
 }
 
 outcome::result<void> ServiceDeployManager::InstallOrbitServicePackage() {
+  CHECK(QThread::currentThread() == thread());
```
Comment (pierricgimmig) [inline OrbitQt/servicedeploymanager.cpp:385]: How does this work? From Qt: currentThread "Returns a pointer to a QThread which manages the currently executing thread." But what is thread() returning?

File: OrbitQt/servicedeploymanager.cpp [hunk 7]
```diff
@@ -67,16 +86,36 @@ ServiceDeployManager::ServiceDeployManager(const DeploymentConfiguration* deploy
...
-      grpc_port_(grpc_port) {
+      grpc_port_(grpc_port),
+      ssh_watchdog_timer_(this) {
   CHECK(deployment_configuration != nullptr);
   CHECK(context != nullptr);
+
+  background_thread_.start();
+  moveToThread(&background_thread_);
+
+  // Event loop needs to be initialized in background thread to work properly on Linux
```
Comment (pierricgimmig) [inline OrbitQt/servicedeploymanager.cpp:97]: mega nit. Missing "."

**Review by pierricgimmig — APPROVED** [review]

File: OrbitQt/servicedeploymanager.cpp [hunk 8]
```diff
@@ -176,9 +220,12 @@ outcome::result<void> ServiceDeployManager::StopSftpChannel(EventLoop* loop,
   return outcome::success();
 }
 
-void ServiceDeployManager::StopSftpChannel() { (void)StopSftpChannel(&loop_, sftp_channel_.get()); }
+void ServiceDeployManager::StopSftpChannel() {
+  (void)StopSftpChannel(&loop_.value(), sftp_channel_.get());
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:224]: Yes, it is. The return type itself is markes as nodiscard.

File: OrbitQt/servicedeploymanager.cpp (hunk 6 above)
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:385]: `thread()` is `this->thread()`, so it calls `QObject::thread` [details](https://doc.qt.io/qt-5/qobject.html#thread)

Every QObject has an associated thread (which can be read by QThread::thread and changed by QObject::moveToThread). Whenever this object receives an event, the event dispatcher will determine the associated thread and queue an event handler task (a slot execution for example) on this thread's event loop.

File: OrbitQt/servicedeploymanager.cpp (hunk 7 above)
Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:97]: Thanks. Easy to fix!

Comment (beckerhe) [reply OrbitQt/servicedeploymanager.cpp:97]: Done.

File: OrbitQt/servicedeploymanager.cpp (hunk 6 above)
Comment (pierricgimmig) [reply OrbitQt/servicedeploymanager.cpp:385]: Ok, thanks for the info!

**Pull request merged**